Add a specific config struct for the update command (#18957)

vdemeester opened the pull request:
This is a proposal for #18957, to reduce what we send to the `update` endpoint. This also make it clear what is mutable in the container, and untie it from `Config`/`HostConfig`. It is still easy to add more attributes

This allows use to control more what is or is not mutable in a container and remove the use of the internal HostConfig struct to be used 🐳.
- [X] `UpdateConfig` might not be a good name, maybe `MutableConfig` or something ?
- [X] The implementation is _naive_ right now, it's there for discussion :wink:.

Related PR for api changes : docker/engine-api#11. Waiting for it to be merged to vendor it and rebase this one :stuck_out_tongue_closed_eyes:.

Closes #18957.

Signed-off-by: Vincent Demeester vincent@sbr.pm

MHBauer commented:
I think **update** as a command is kind of a goofy name for the command, but I wasn't here a week ago to say that. I'd match the config to the name of the command. If the command is `update`, the config should be `update`.

hqhq commented:
Design looks good, I'm +1 for this :+1:

duglin commented:
SGTM

MHBauer commented on api/server/router/container/container_routes.go:
```diff
@@ -332,11 +333,17 @@ func (s *containerRouter) postContainerUpdate(ctx context.Context, w http.Respon
 		return err
 	}
 
+	hostConfig := &container.HostConfig{
```
Is there further work to do here to get rid of hostConfig? If we end up in a hostConfig anyway, are we really winning anything here?

hqhq replied:
`hostConfig` here is invisible to users, I think that's the point.

vdemeester replied:
Yeah I didn't want to pass `UpdateConfig` down too much — I think `daemon` doesn't really need to know it.

MHBauer commented:
Looks correct.

calavera commented:
it looks like #19116 is proving that keeping hostConfig is better than removing it completely.

Tbh, I was going to agree with this change, but allowing to change the restartPolicy sounds like a good idea to me. So I'm leaning more towards not doing this.

vdemeester commented:
@calavera Adding support to RestartPolicy would be done to just add a new field to `UpdateConfig`. What I don't like now is that we pass the whole HostConfig and ignoring a whole lot of attributes (for good reasons)… but seeing the API, a consumer of this API could think "oh cool, I can update everything in `HostConfig`", which is not the case, and shouldn't be.

tiborvass commented:
*Collective review*

We're okay with this design to resolve #18957.

Thanks @vdemeester

tiborvass commented:
@vdemeester rebase please :)

vdemeester commented:
Rebased and added a probably temporary vendoring commit (cc @calavera) to make it build :stuck_out_tongue_closed_eyes:.

vdemeester commented:
Rebased with the vendoring of `engine-api` 0.1.2 :wink:

duglin commented:
So weird to only see a small piece of the puzzle  :-)
LGTM

tiborvass commented:
LGTM

tiborvass commented:
@vdemeester not sure if this needs a documentation update too, or if that should be done somewhere else?

vdemeester commented:
Hum maybe remote api I think, I'll check once home ;-)

vdemeester commented:
Updated some API docs.

jamtur01 commented:
Docs LGTM

jamtur01 commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -95,6 +95,7 @@ This section lists each version from latest to oldest.  Each listing includes a
 
 [Docker Remote API v1.22](docker_remote_api_v1.22.md) documentation
 
+* `POST /container/(name)/update` updates resources of a container.
```
updates the resources

vdemeester committed:
Add a specific config for the update command

This allows to define clearly what is mutable or not in a container
and remove the use of the internal HostConfig struct to be used.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
docs LGTM, thanks @vdemeester!
